### PR TITLE
Adding ros_diagnostics publication into rosmon_core.

### PR DESCRIPTION
**Pull request by adrienbarral**
Here is my pull request for the feature that add ros diagnostics into rosmon core instead of adding it as a third part node.

### CODE REVIEW

File: rosmon_core/test/memory_parser.test.cc [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#include <gtest/gtest.h>
```
Comment (xqms) [inline rosmon_core/test/memory_parser.test.cc]: I much prefer `catch_ros` as the unit testing framework, as this is already used everywhere else in rosmon. If you are not familiar with it, I can convert your testcase later on, no problem.

File: rosmon_core/src/launch/bytes_parser.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#include <fmt/format.h>
+#include <ros/console.h>
+
+namespace rosmon {
+
+namespace launch {
+class ByteParser
+{
+public:
+    bool parseMemory(const std::string& memory, uint64_t& outMemory)
```
[the remark this answers is not in the document]
Comment (xqms) [reply rosmon_core/src/launch/bytes_parser.h]: ... and then we could hide the implementation and all boost::spirit includes in a `byte_parser.cpp`.

[the remark this answers is not in the document]
Comment (adrienbarral) [reply rosmon_core/src/launch/bytes_parser.h]: Commit 59a92dc1d6a7d5090c7a676cfabc6a601afe073b solve this.

File: rosmon_core/src/launch/bytes_parser.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            add("kB", 1e3)("mB", 1e6)("gB", 1e9)("tB", 1e12)("KB", 1e3)("MB", 1e6)(
+                "GB", 1e9)("TB", 1e12)("B", 1);
+        }
+    } byte_decades;
+};
+
+class ByteStringGenerator {
+public :
+    static std::string memoryToString(uint64_t memory)
```
[the remark this answers is not in the document]
Comment (adrienbarral) [reply rosmon_core/src/launch/bytes_parser.h]: Fixed in 57b927d64f043ac6e5e0854d1a8709dde8a394d0

File: rosmon_core/src/diagnostics/rosmon_to_diagnostic.cpp [hunk 4]
```diff
@@ -73,3 +75,24 @@ void RosmonToDiagnostic::updateDiagnostics(const std::vector<NodeMonitor::Ptr>&
...
+
+std::string RosmonToDiagnostic::memoryToString(uint64_t memory)
+{
+    if(memory < 1000)
+    {
+        return fmt::format("{} B", memory);
+    }
+    else if(memory < 1e6)
+    {
+        return fmt::format("{:.2f} KiB", memory / 1e3);
```
Comment (xqms) [inline rosmon_core/src/diagnostics/rosmon_to_diagnostic.cpp]: If we are talking `KiB`, you should check against 1024*1024 and divide by 1024, not 1000. Sorry, should have mentioned that.

**Comment by xqms** [comment]
Thanks for the update :+1:
I see two things to do before this can be merged:

- [ ] Investigate CI failures - it seems that there is a missing dependency on diagnostics_msgs or similar.
- [ ] Port unit tests to `catch_ros`. I can look into this, but probably not in the next week.

**Comment by adrienbarral** [comment]
I can't really understand why the CI test fail on jenkins. we have the following error on the ros build farm : 

```
21:28:24 Scanning dependencies of target _run_tests_rosmon_core_rostest_test_basic.test
21:28:24 -- run_tests.py: execute commands
21:28:24   /opt/ros/lunar/share/rostest/cmake/../../../bin/rostest --pkgdir=/tmp/ws/src/rosmon/rosmon_core --package=rosmon_core --results-filename test_basic.xml --results-base-dir "/tmp/ws/test_results" /tmp/ws/src/rosmon/rosmon_core/test/basic.test 
21:28:25 ... logging to /home/buildfarm/.ros/log/rostest-408a64e457cb-1910.log
21:28:26 [ROSUNIT] Outputting test results to /tmp/ws/test_results/rosmon_core/rostest-test_basic.xml
21:28:28 terminate called after throwing an instance of 'fmt::v5::format_error'
21:28:28   what():  precision not allowed for this argument type
```

On my computer when I try to run this test I have another error !

```
Traceback (most recent call last):
  File "/home/aba/Projects/MON/workspace/src/rosmon/rosmon_core/test/basic.py", line 20, in <module>
    from rosmon_msgs.msg import State, NodeState
ImportError: No module named msg
          nested_mon: nested_mon exited with status 0
               test1: test1 exited with status 0
               test2: test2 exited with status 0
[Testcase: testrosmon_test] ... FAILURE!
```

Do you have any idea ?

**Comment by xqms** [comment]
On my system the test fails with exactly the same exception as on the CI server. Maybe you did not build `rosmon_msgs` or did not source the workspace devel file?

To replicate & debug, you can do:
```
cd rosmon_core/test
gdb --args ../../../../devel/lib/rosmon_core/rosmon --disable-ui rosmon_core basic.launch
```

Full backtrace:
```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff53c3801 in __GI_abort () at abort.c:79
#2  0x00007ffff5db68b7 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff5dbca06 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff5dbca41 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff5dbcc74 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff7bb4045 in fmt::v5::internal::error_handler::on_error (this=<optimized out>, message=0x5555555b6b88 "precision not allowed for this argument type") at /opt/ros/melodic/include/fmt/format-inl.h:821
#7  0x0000555555593fc3 in fmt::v5::basic_parse_context<char, fmt::v5::internal::error_handler>::on_error (message=0x5555555b6b88 "precision not allowed for this argument type", this=<optimized out>) at /home/max/projects/release/devel/.private/rosfmt/include/fmt/core.h:863
#8  fmt::v5::internal::context_base<std::back_insert_iterator<fmt::v5::internal::basic_buffer<char> >, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<char> >, char>, char>::on_error (
    message=0x5555555b6b88 "precision not allowed for this argument type", this=<optimized out>) at /home/max/projects/release/devel/.private/rosfmt/include/fmt/core.h:953
#9  fmt::v5::internal::specs_handler<fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<char> >, char> >::on_error (this=0x7fffffffba30, message=0x5555555b6b88 "precision not allowed for this argument type")
    at /home/max/projects/release/devel/.private/rosfmt/include/fmt/format.h:1823
#10 fmt::v5::internal::specs_checker<fmt::v5::internal::specs_handler<fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<char> >, char> > >::end_precision (this=0x7fffffffba30) at /home/max/projects/release/devel/.private/rosfmt/include/fmt/format.h:1768
#11 fmt::v5::internal::parse_format_specs<fmt::v5::internal::null_terminating_iterator<char>, fmt::v5::internal::specs_checker<fmt::v5::internal::specs_handler<fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<char> >, char> > >&> (handler=..., it=...)
    at /home/max/projects/release/devel/.private/rosfmt/include/fmt/format.h:2107
#12 fmt::v5::format_handler<fmt::v5::arg_formatter<fmt::v5::back_insert_range<fmt::v5::internal::basic_buffer<char> > >, char, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<char> >, char> >::on_format_specs (it=..., this=0x7fffffffbac0)
    at /home/max/projects/release/devel/.private/rosfmt/include/fmt/format.h:3328
#13 fmt::v5::internal::parse_format_string<false, char, fmt::v5::format_handler<fmt::v5::arg_formatter<fmt::v5::back_insert_range<fmt::v5::internal::basic_buffer<char> > >, char, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<char> >, char> >&> (
    format_str=..., handler=...) at /home/max/projects/release/devel/.private/rosfmt/include/fmt/format.h:2188
#14 0x000055555559f7b6 in fmt::v5::vformat_to<fmt::v5::arg_formatter<fmt::v5::back_insert_range<fmt::v5::internal::basic_buffer<char> > >, char, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<char> >, char> > (args=..., format_str=..., out=...)
    at /home/max/projects/release/devel/.private/rosfmt/include/fmt/format.h:3346
#15 fmt::v5::vformat_to (args=..., format_str=..., buf=Warnung: RTTI symbol not found for class 'fmt::v5::basic_memory_buffer<char, 500ul, std::allocator<char> >'
...) at /home/max/projects/release/devel/.private/rosfmt/include/fmt/format.h:3451
#16 fmt::v5::internal::vformat<char> (format_str=..., args=...) at /home/max/projects/release/devel/.private/rosfmt/include/fmt/format.h:3574
#17 0x00005555555a76d9 in fmt::v5::format<char [5], unsigned long> (format_str=...) at /home/max/projects/release/devel/.private/rosfmt/include/fmt/core.h:1446
#18 rosmon::diagnostics::RosmonToDiagnostic::memoryToString[abi:cxx11](unsigned long) (memory=<optimized out>) at /home/max/projects/release/src/rosmon/rosmon_core/src/diagnostics/rosmon_to_diagnostic.cpp:91
#19 0x00005555555a7cbf in rosmon::diagnostics::RosmonToDiagnostic::updateDiagnostics (this=0x555555ccbd60, state=...) at /home/max/projects/release/src/rosmon/rosmon_core/src/diagnostics/rosmon_to_diagnostic.cpp:41
#20 0x00005555555ae0e0 in rosmon::ROSInterface::update (this=0x7fffffffceb0) at /home/max/projects/release/src/rosmon/rosmon_core/src/ros_interface.cpp:38
#21 0x00007ffff78c4e17 in ros::TimerManager<ros::WallTime, ros::WallDuration, ros::WallTimerEvent>::TimerQueueCallback::call() () from /opt/ros/melodic/lib/libroscpp.so
#22 0x00007ffff78f8f1c in ros::CallbackQueue::callOneCB(ros::CallbackQueue::TLS*) () from /opt/ros/melodic/lib/libroscpp.so
#23 0x00007ffff78fa30b in ros::CallbackQueue::callAvailable(ros::WallDuration) () from /opt/ros/melodic/lib/libroscpp.so
#24 0x00007ffff793a4ce in ros::spinOnce() () from /opt/ros/melodic/lib/libroscpp.so
#25 0x0000555555581cbf in main (argc=<optimized out>, argv=<optimized out>) at /home/max/projects/release/src/rosmon/rosmon_core/src/main.cpp:468
```

It's failing here:
```
#18 rosmon::diagnostics::RosmonToDiagnostic::memoryToString[abi:cxx11](unsigned long) (memory=<optimized out>) at /home/max/projects/release/src/rosmon/rosmon_core/src/diagnostics/rosmon_to_diagnostic.cpp:91
91              return fmt::format("{:.2f} MiB", memory / static_cast<uint64_t>(1<<20));
```

I guess the `.2f` modifier only works for float, but `memory / ...` is an integer.

**Comment by xqms** [comment]
Seems to work in the workspaces I have for testing. I raised the default memory & CPU limits though, many nodes use more than 50MB memory and that's not a problem on most platforms.

I'll wait for CI results and then we can merge :-)

**Comment by xqms** [comment]
Seems to work in the workspaces I use for testing. I raised the default memory & CPU limits though, there are were many nodes using more than 50MB RAM where I would say that is not a problem.

Let's go ahead and merge this! :-)
Since this is kind of an additional feature, I'm fine with testing it more in master.

**Comment by xqms** [comment]
huh, strange issue with the comment timestamps, sorry for the double post. Somehow I'm able to write from the future...

**Pull request merged**